File: android/app/src/main/java/com/soundcam/camera/net/CommandClient.kt
package com.soundcam.camera.net

import android.util.Log
import com.soundcam.camera.protocol.DeviceInfo
import com.soundcam.camera.protocol.Messages
import com.soundcam.camera.protocol.RemoteSettings
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Client WebSocket persistant du canal de commandes (EF-21).
 *
 * Le téléphone est client ; le PC serveur. Reconnexion automatique avec backoff
 * exponentiel (spéc. §7.1). À la connexion, envoie un HELLO d'appairage (ENF-06) ;
 * sur HELLO_ACK accepté, mémorise le token fort via [callback].
 */
class CommandClient(
    private val device: DeviceInfo,
    private val callback: Callback,
) {

    interface Callback {
        fun onConnected(serverName: String?)
        fun onPaired(token: String)
        fun onPairingRejected(reason: String?)
        fun onStartRecording(requestId: String, maxDurationS: Double)
        fun onStopRecording(requestId: String)
        fun onSetSettings(requestId: String, settings: RemoteSettings)
        fun onLiveRequest(sessionId: String)
        fun onLiveAnswer(sessionId: String, sdp: String)
        fun onLiveIce(sessionId: String, candidate: String, sdpMid: String?, sdpMLineIndex: Int)
        fun onLiveStop(sessionId: String)
        fun onDisconnected()
    }

    @Volatile private var webSocket: WebSocket? = null
    @Volatile private var credential: String = ""
    @Volatile private var url: String = ""
    @Volatile private var tls: Boolean = false
    private val running = AtomicBoolean(false)
    private var backoffMs = INITIAL_BACKOFF_MS
    private val client: OkHttpClient by lazy {
        TlsTrust.client(
            tls,
            OkHttpClient.Builder()
                .pingInterval(5, TimeUnit.SECONDS)
                .connectTimeout(8, TimeUnit.SECONDS)
                .readTimeout(0, TimeUnit.MILLISECONDS),
        )
    }

    val isConnected: Boolean get() = webSocket != null

    /** Démarre (ou met à jour) la connexion vers le serveur découvert/configuré. */
    fun connect(host: String, port: Int, tls: Boolean, credential: String) {
        this.url = (if (tls) "wss" else "ws") + "://$host:$port/ws"
        this.tls = tls
        this.credential = credential
        if (running.compareAndSet(false, true)) {
            openSocket()
        } else {
            // Reconfiguration : ferme la socket courante pour rouvrir avec la nouvelle URL.
            webSocket?.cancel()
        }
    }

    private fun openSocket() {
        if (!running.get() || url.isEmpty()) return
        Log.i(TAG, "Connexion à $url")
        val request = Request.Builder().url(url).build()
        webSocket = client.newWebSocket(request, Listener())
    }

    private inner class Listener : WebSocketListener() {
        override fun onOpen(ws: WebSocket, response: Response) {
            ws.send(Messages.hello(credential, device))
        }

        override fun onMessage(ws: WebSocket, text: String) {
            handle(text)
        }

        override fun onClosing(ws: WebSocket, code: Int, reason: String) {
            ws.close(1000, null)
        }

        override fun onClosed(ws: WebSocket, code: Int, reason: String) {
            onSocketDown()
        }

        override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
            Log.w(TAG, "WebSocket échec: ${t.message}")
            onSocketDown()
        }
    }

    private fun handle(text: String) {
        val o: JSONObject = runCatching { Messages.parse(text) }.getOrNull() ?: return
        when (o.optString("type")) {
            Messages.HELLO_ACK -> {
                if (o.optBoolean("accepted")) {
                    backoffMs = INITIAL_BACKOFF_MS
                    callback.onConnected(o.optString("server_name").ifEmpty { null })
                    val token = o.optString("token", "")
                    if (token.isNotEmpty()) {
                        credential = token
                        callback.onPaired(token)
                    }
                } else {
                    callback.onPairingRejected(o.optString("reason").ifEmpty { null })
                    running.set(false) // appairage refusé : ne pas boucler indéfiniment
                }
            }
            Messages.START_RECORDING ->
                callback.onStartRecording(o.optString("request_id"), o.optDouble("max_duration_s", 0.0))
            Messages.STOP_RECORDING ->
                callback.onStopRecording(o.optString("request_id"))
            Messages.SET_SETTINGS ->
                callback.onSetSettings(o.optString("request_id"), RemoteSettings.fromJson(o))
            Messages.LIVE_REQUEST ->
                callback.onLiveRequest(o.optString("session_id"))
            Messages.LIVE_ANSWER ->
                callback.onLiveAnswer(o.optString("session_id"), o.optString("sdp"))
            Messages.LIVE_ICE -> {
                val c = o.optJSONObject("candidate")
                if (c != null) callback.onLiveIce(
                    o.optString("session_id"),
                    c.optString("candidate"),
                    if (c.isNull("sdpMid")) null else c.optString("sdpMid"),
                    c.optInt("sdpMLineIndex", 0),
                )
            }
            Messages.LIVE_STOP ->
                callback.onLiveStop(o.optString("session_id"))
        }
    }

    private fun onSocketDown() {
        webSocket = null
        callback.onDisconnected()
        if (running.get()) scheduleReconnect()
    }

    private fun scheduleReconnect() {
        val delay = backoffMs
        backoffMs = (backoffMs * 2).coerceAtMost(MAX_BACKOFF_MS)
        Thread {
            Thread.sleep(delay)
            if (running.get()) openSocket()
        }.start()
    }

    fun sendStatus(state: String, recording: Boolean, battery: Float?, recordingId: String?) {
        webSocket?.send(Messages.status(state, recording, battery, recordingId))
    }

    fun sendAck(requestId: String, result: String, reason: String?, recordingId: String?) {
        webSocket?.send(Messages.ack(requestId, result, reason, recordingId))
    }

    fun sendSoundTriggered(recordingId: String, confidence: Float, label: String?) {
        webSocket?.send(Messages.soundTriggered(recordingId, confidence, label))
    }

    fun sendLiveOffer(sessionId: String, sdp: String) {
        webSocket?.send(Messages.liveOffer(sessionId, sdp))
    }

    fun sendLiveIce(sessionId: String, candidate: String, sdpMid: String?, sdpMLineIndex: Int) {
        webSocket?.send(Messages.liveIce(sessionId, candidate, sdpMid, sdpMLineIndex))
    }

    fun sendLiveStop(sessionId: String) {
        webSocket?.send(Messages.liveStop(sessionId))
    }

    fun close() {
        running.set(false)
        webSocket?.close(1000, "bye")
        webSocket = null
    }

    companion object {
        private const val TAG = "CommandClient"
        private const val INITIAL_BACKOFF_MS = 1000L
        private const val MAX_BACKOFF_MS = 30_000L
    }
}

File: ios/SoundCam/Net/CommandClient.swift
import Foundation

/// Client WebSocket persistant du canal de commandes (EF-21), basé sur
/// `URLSessionWebSocketTask`. Reconnexion automatique avec backoff exponentiel.
///
/// En mode TLS, le PC présente un certificat auto-signé local (ENF-07) ; on l'accepte
/// après appairage explicite (modèle TOFU, voir `URLSessionDelegate`).
final class CommandClient: NSObject, URLSessionDelegate {

    protocol Delegate: AnyObject {
        func onConnected(serverName: String?)
        func onPaired(token: String)
        func onPairingRejected(reason: String?)
        func onStartRecording(requestId: String, maxDurationS: Double)
        func onStopRecording(requestId: String)
        func onSetSettings(requestId: String, settings: RemoteSettings)
        func onDisconnected()
    }

    weak var delegate: Delegate?
    private let device: DeviceInfo
    private var task: URLSessionWebSocketTask?
    private lazy var session: URLSession = {
        URLSession(configuration: .default, delegate: self, delegateQueue: nil)
    }()

    private var credential = ""
    private var urlString = ""
    private var running = false
    private var backoff: TimeInterval = 1.0
    private let maxBackoff: TimeInterval = 30.0

    private(set) var isConnected = false

    init(device: DeviceInfo) {
        self.device = device
    }

    func connect(host: String, port: Int, tls: Bool, credential: String) {
        let scheme = tls ? "wss" : "ws"
        urlString = "\(scheme)://\(host):\(port)/ws"
        self.credential = credential
        running = true
        openSocket()
    }

    private func openSocket() {
        guard running, let url = URL(string: urlString) else { return }
        let task = session.webSocketTask(with: url)
        self.task = task
        task.resume()
        send(Messages.hello(token: credential, device: device))
        receiveLoop()
    }

    private func receiveLoop() {
        task?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .failure:
                self.handleDisconnect()
            case .success(let message):
                if case let .string(text) = message {
                    self.handle(text)
                }
                self.receiveLoop()
            }
        }
    }

    private func handle(_ text: String) {
        guard let o = Messages.decode(text), let type = o["type"] as? String else { return }
        switch type {
        case Messages.MsgType.helloAck:
            let accepted = (o["accepted"] as? Bool) ?? false
            if accepted {
                isConnected = true
                backoff = 1.0
                delegate?.onConnected(serverName: o["server_name"] as? String)
                if let token = o["token"] as? String, !token.isEmpty {
                    credential = token
                    delegate?.onPaired(token: token)
                }
            } else {
                running = false
                delegate?.onPairingRejected(reason: o["reason"] as? String)
            }
        case Messages.MsgType.startRecording:
            delegate?.onStartRecording(
                requestId: o["request_id"] as? String ?? "",
                maxDurationS: (o["max_duration_s"] as? NSNumber)?.doubleValue ?? 0)
        case Messages.MsgType.stopRecording:
            delegate?.onStopRecording(requestId: o["request_id"] as? String ?? "")
        case Messages.MsgType.setSettings:
            delegate?.onSetSettings(
                requestId: o["request_id"] as? String ?? "",
                settings: RemoteSettings(from: o))
        default:
            break
        }
    }

    private func handleDisconnect() {
        isConnected = false
        task = nil
        delegate?.onDisconnected()
        guard running else { return }
        let delay = backoff
        backoff = min(backoff * 2, maxBackoff)
        DispatchQueue.global().asyncAfter(deadline: .now() + delay) { [weak self] in
            self?.openSocket()
        }
    }

    func send(_ text: String) {
        task?.send(.string(text)) { _ in }
    }

    func sendStatus(state: String, recording: Bool, battery: Float?, recordingId: String?) {
        send(Messages.status(state: state, recording: recording, battery: battery, recordingId: recordingId))
    }

    func sendAck(requestId: String, result: String, reason: String?, recordingId: String?) {
        send(Messages.ack(requestId: requestId, result: result, reason: reason, recordingId: recordingId))
    }

    func sendSoundTriggered(recordingId: String, confidence: Float, label: String?) {
        send(Messages.soundTriggered(recordingId: recordingId, confidence: confidence, label: label))
    }

    func close() {
        running = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    // TOFU : accepte le certificat auto-signé du PC sur le réseau local (ENF-07).
    func urlSession(_ session: URLSession,
                    didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        if let trust = challenge.protectionSpace.serverTrust {
            completionHandler(.useCredential, URLCredential(trust: trust))
        } else {
            completionHandler(.performDefaultHandling, nil)
        }
    }
}
